R/bernoulli.R

```diff
@@ -157,3 +157,53 @@ cdf.bernoulli <- function(d, x, ...) {
 quantile.bernoulli <- function(d, p, ...) {
   qbinom(p = p, size = 1, prob = d$p)
 }
+
+#' Fit a bernoulli distribution to data
+#'
+#' @param d A `bernoulli` object.
+#' @param x A vector of zeroes and ones
+#'
+#' @return a `bernoulli` object
+#' @export
+fit_mle.bernoulli <- function(d, x, ...) {
+  ss <- suff_stat(d, x, ...)
+  bernoulli(p = ss$successes / (ss$successes + ss$failures))
+}
+
+#' Compute the sufficient statistics for a bernoulli distribution from data
+#'
+#' @inherit beroulli
+#'
+#' @export
+suff_stat.bernoulli <- function(d, x, ...) {
+  valid_x <- (x %in% c(0L, 1L))
+  if(any(!valid_x)) stop("`x` contains elements other than 0 or 1")
+  list(successes = sum(x == 1), failures = sum(x == 0))
+}
+
+#' Compute the likelihood of a bernoulli distribution given data
+#'
+#' @inheritParams fit_mle.bernoulli
+#'
+#' @return the likelihood
+#' @export
+
+likelihood.bernoulli <- function(d, x, ...) {
+  prod(dbinom(x = x, size = 1, prob = d$p))
+}
+
+#' Compute the log-likelihood of a bernoulli distribution given data
+#'
+#' @inheritParams fit_mle.bernoulli
+#'
+#' @return the log-likelihood
+#' @export
+log_likelihood.bernoulli <- function(d, x, ...) {
+  sum(dbinom(x = x, size = 1, prob = d$p, log = TRUE))
```

Comment:
I like this but think we should use the generic based approach here. Which means implementing a bunch of `log_pdf()` methods ugh. If you're willing to do this, great. If not, I'll merge and take care of it down the line.

Reply:
Yeah, actually a `log_pdf` generic would solve for `log_likelihood` and in turn `likelihood`. That would also mean that potentially we wouldn't need generics for them? Just have something like 

```r
log_likelihood <- function(d, x) {
  sum(log_pdf(d, x))
}

likelihood <- function(d, x) {
  exp(log_likelihood(d, x))
}
```

Reply:
I was wondering if just using `log(pdf(d, x))` was a good idea over implementing an entirely new generic that would be nearly identical but it mentions in the docs that the `log = TRUE` argument in all the base distributions has a greater range.

```r
n <- 2000
k <- seq(0, n, by = 20)
plot (k, dbinom(k, n, pi/10, log = TRUE), type = "l", ylab = "log density",
      main = "dbinom(*, log=TRUE) is better than  log(dbinom(*))")
lines(k, log(dbinom(k, n, pi/10)), col = "red", lwd = 2)
```

So it's worth it really

Reply:
Yeah definitely worth it for numerical reasons. Taking the log of tiny numbers is sketchy and I'm sure the `log = TRUE` option does smart stuff.

Reply:
So I've been thinking about this over the last few days and am just coming into some time to work on it, but wanted to get your thoughts. I think ultimately there's 3 options, and each has their pros and cons and wanted to put them past you. 

## Option 1

Create a `log_pdf` generic and write methods for every distribution. This is straight forward but results in a lot of code duplication and makes updating and adding code down the line slightly more difficult. For example

```r
pdf.binomial <- function(d, x, ...) {
  dbinom(x = x, size = d$size, prob = d$p)
}

log_pdf.binomial <- function(d, x, ...) {
  dbinom(x = x, size = d$size, prob = d$p, log = TRUE)
}
``` 

This seems slightly ridiculous when you look at it.

## Option 2 

Add a `log` argument to `pdf` and just create a function `log_pdf` that uses this

```r
pdf.binomial <- function(d, x, log = FALSE, ...) {
  dbinom(x = x, size = d$size, prob = d$p, log = log)
}

log_pdf <- function(d, x, ...) {
  pdf(d, x, log = TRUE, ...)
}
```

This cuts down on duplication considerably and would be very easy to update existing distributions for. It also falls inline with the APIs for the base R distributions. 


## Option 3

Option 3 is to make use of dot-dot-dot, but I never really use dot-dot-dot and am not sure of the implications if any

```r
pdf.binomial <- function(d, x, ...) {
  dbinom(x = x, size = d$size, prob = d$p, ...)
}

log_pdf <- function(d, x, ...) {
  pdf(d, x, log = TRUE, ...)
}
```

Option 1 gives them most control, because we can write an exact `log_pdf` function for every distribution. I'm not sure if there is much benefit in that, but maybe there are cases where this is preferable? 

Option 2 and 3 are simple which is nice. Especially given really we just want to take the log of a probability. A downside is that every `pdf` method needs to handle the log implementation internally, but that is already done for the base R functions. 

My preference would be Option 2 because it's the most explicit. Any thoughts yourself?

Reply:
I am quite strongly in favor of Option 1!

Reply:
Grand, consider it done

Reply:
Right they're all written - documentation just points to the `pdf` method for each distribution and I've added a `log_pdf` example for each too. No tests included though. I can start building them out next.